### DISCUSSION TITLE
Ribbon

### DISCUSSION THREAD

**Discussion by vgromfeld** [CommunityToolkit/Labs-Windows#544] [Experiments]
## Problem statement
Office and applications like Paint are using a "large" ribbon to display all their available commands. Those ribbons display all the available commands in groups. Those groups can automatically be collapsed if there is not enough space available to display them all based on a priority order.

The toolkit already provides a "small" ribbon through the `TabbedCommandBar` (See #500 and [Porting TabbedCommandBar #243](https://github.com/CommunityToolkit/Windows/pull/243)) but does not have any way to display the "large" ribbon.

### Default
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/2a72ea26-110c-4396-874b-b985804c58fa)
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/9e0c4aa8-0aed-4f3f-a099-6c18010ca858)

### Narrow
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/23977a81-ec58-45ab-bb47-711f23f7f7bd)
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/649056d2-4e53-4a23-b89c-af2756982cb5)

## Proposed solution
Add a new custom control to the toolkit.
It will use a custom layout to automatically collapse the ribbon's groups when there is not enough space available.
If all the groups cannot fit in the available width, the ribbon should display a left and/or right scroll buttons.

```xml
<Ribbon>
   <RibbonGroup Label="Selection" />
    <RibbonCollapsibleGroup Label="Edit" Style="{StaticResource RibbonLeftCollapsibleGroupStyle}" Priority="1">
        <!-- group content -->
    </RibbonCollapsibleGroup>
    <RibbonCollapsibleGroup Label="Shapes" Style="{StaticResource RibbonLeftCollapsibleGroupStyle}" Priority="2">
        <!-- group content -->
    </RibbonCollapsibleGroup>
</Ribbon>
```

### Result
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/b2c98722-7bf3-4fd8-9d9b-d23443e11f7e)
![image](https://github.com/CommunityToolkit/Labs-Windows/assets/20152272/80d5db8c-ed2c-43f7-9725-c058ebf7788b)

**Comment by michael-hawker** [comment]
For future reference, the [WPF Ribbon docs are here](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.ribbon.ribbon).

One main difference here is that this is just the internals of a Ribbon tab, vs. the WPF control which handled the tabs itself.

However, now with both `SelectorBar` and the Toolkit's `TabbedCommandBar`, this solution could probably be combined more modularly with those vs. trying to handle everything together, eh? Though don't think we've tried to combine the later yet.

It should be something we think about for a Sample though in order to show the complete experience.

**Comment by Arlodotexe** [comment]
This was added in https://github.com/CommunityToolkit/Labs-Windows/pull/546. See original [duplicate] discussion in https://github.com/CommunityToolkit/Labs-Windows/discussions/500.

**Comment by DL444** [comment]
I see the currently available API is simply `RibbonGroup` hosting arbitrary content. Should we in the future expose common group patterns at a higher abstraction level? For example, implementing adaptive command groups (see Layout - Page Setup in Microsoft Word) where commands are labeled and stacked optionally based on available width can be pretty complicated with today's primitive API (`RequestedWidths`), especially if internationalization requires width stops to be computed dynamically at run time.

I further imagine future complex scenarios that could benefit from a close integration between `Ribbon`, `RibbonGroup`, and their content. Having group encapsulations higher up the stack enables this integration.
